# LDlinkR <img src="man/figures/logo.png" align="right" height=140/>

<!-- badges: start -->

[![CRAN
version](https://www.r-pkg.org/badges/version-ago/LDlinkR)](https://cran.r-project.org/package=LDlinkR)
[![CRAN RStudio mirror
downloads](https://cranlogs.r-pkg.org/badges/grand-total/LDlinkR?color=blue)](https://r-pkg.org/pkg/LDlinkR)
![CRAN/METACRAN](https://img.shields.io/cran/l/LDlinkR?color=yellow)


### An R Package for Calculating Linkage Disequilibrium

## Description

[LDlink](https://ldlink.nci.nih.gov) is an interactive and powerful suite of web-based tools for querying germline variants in human population groups of interest to generate interactive tables and plots.  All population genotype data originates from Phase 3 (Version 5) of the [1000 Genomes Project](https://www.internationalgenome.org/) and variant RS (reference SNP) numbers are indexed based on dbSNP build 151.

*LDlinkR* is an R package developed to query and download results generated by [LDlink](https://ldlink.nci.nih.gov) web-based applications from the R console.  It facilitates researchers who are interested in performing batch queries.  *LDlinkR* accelerates genomic research by providing efficient and user-friendly functions to programmatically interrogate pairwise linkage disequilibrium from large lists of genetic variants. 

## Installation
- The release version of *LDlinkR* can be installed from [CRAN](https://cran.r-project.org/) with:
``` r
install.packages("LDlinkR")
```
- The development version of the *LDlinkR* package can be installed from the [GitHub repository](https://github.com/CBIIT/LDlinkR) by using the `remotes` package:
``` r 
install.packages("remotes")
remotes::install_github("CBIIT/LDlinkR")
```
*LDlinkR* depends on the following packages:

- *utils*, version 3.4.2 or later
- *httr*, version 1.4.0 or later

Following installation, attach the *LDlinkR* package with:
``` r
library(LDlinkR)
```

## Personal Access Token - **Required**

In order to access the [LDlink](https://ldlink.nci.nih.gov) API via *LDlinkR*, we use a personal access token. This is a common convention followed by many APIs and emulates the more familiar HTTPS username/password or SSH keys.

You will need to:

- Make a one-time request for your personal access token from a web browser at <https://ldlink.nci.nih.gov/?tab=apiaccess>.
- Once registered, your personal access token will be emailed to you. It is a string of 12 random letters and numbers. 
- Provide your token as an argument when using *LDlinkR*.  See example below:

``` r
LDhap(snps = c("rs3", "rs4", "rs148890987"), 
      pop = "YRI", 
      token = "YourTokenHere123")
```

<br>

## Available Functions
| Function| Description|
|:--------|:-----------|
|`LDexpress`|Search if a list of genomic variants is associated with gene expression in tissues of interest.|
|`LDhap` |Calculates population specific haplotype frequencies of all haplotypes observed for a list of query variants.|
|`LDmatrix` |Generates a data frame of pairwise linkage disequilibirum statistics.|
|`LDpair`|Investigates potentially correlated alleles for a pair of variants.|
|`LDpop` |Investigates allele frequencies and linkage disequilibrium patterns across 1000G populations.|
|`LDproxy` |Explore proxy and putatively functional variants for a single query variant.|
|`LDproxy_batch`|Query `LDproxy` using a list of query variants.|
|`SNPchip` |Find commercial genotyping chip arrays for variants of interest.|
|`SNPclip` |Prune a list of variants by linkage disequilibrium.|

<br>

## Utilities
|Utility Function |Description|
|:--------|:-----------|
|`list_chips` |Provides a data frame listing the names and abbreviation codes for available commerical SNP Chip Arrays from Illumina and Affymetrix.|
|`list_pop` |Provides a data frame listing the available reference populations from the [1000 Genomes Project](https://www.internationalgenome.org/).|
|`list_gtex_tissues` |Provides a data frame listing the GTEx full names, `LDexpress` full names (without spaces) and acceptable abbreviation codes of the 54 non-diseased tissue sites collected for the [GTEx Portal](https://gtexportal.org/home/) and used as input for the `LDexpress` function.|


<br>

## Contributors

Timothy A. Myers, Stephen J. Chanock and Mitchell J. Machiela

