<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Timothy A. Myers, Stephen J. Chanock and Mitchell J. Machiela" />

<meta name="date" content="2023-05-30" />

<title>LDlinkR: An R Package for Rapidly Calculating Linkage Disequilibrium Statistics in Diverse Populations</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore"><em>LDlinkR</em>: An R Package for Rapidly
Calculating Linkage Disequilibrium Statistics in Diverse
Populations</h1>
<h4 class="author">Timothy A. Myers, Stephen J. Chanock and Mitchell J.
Machiela</h4>
<h4 class="date">2023-05-30</h4>



<hr />
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p><a href="https://ldlink.nih.gov">LDlink</a> is an interactive and
powerful suite of web-based tools for querying germline variants in
human population groups of interest to generate interactive tables and
plots. All population genotype data originates from Phase 3 (Version 5)
of the 1000 Genomes Project and variant RS numbers are indexed based on
<a href="https://www.ncbi.nlm.nih.gov/snp/">dbSNP</a> 155.</p>
<p><em>LDlinkR</em> is an R package developed to query and download
results <strong>(internet access required)</strong> generated by LDlink
web-based applications from the R console. <em>LDlinkR</em> accelerates
genomic research by providing efficient and user-friendly functions to
programmatically interrogate pairwise linkage disequilibrium from large
lists of genetic variants.</p>
<p>Please see the the sections below and the online <a href="https://ldlink.nih.gov/?tab=help#Understanding_Linkage_Disequilibrium">LDlink
documentation</a> for more information about understanding linkage
disequilibrium (LD) and additional details about how LDlink calculates
patterns of LD across a variety of ancestral human populations.</p>
</div>
<div id="understanding-linkage-disequilibrium" class="section level2">
<h2>Understanding Linkage Disequilibrium</h2>
<p>What is linkage <strong>dis</strong>equilibrium? Perhaps it is best
to start with linkage <strong>equilibrium</strong>. Linkage equilibrium
exists when alleles from two different genetic variants occur
independently of each other. The inheritance of such variants follows
probabilistic patterns governed by population allele frequencies. The
vast majority of genetic variants on a chromosome are in linkage
equilibrium. Variants in linkage equilibrium are not considered
linked.</p>
<p>Linkage <strong>dis</strong>equilibrium is present when alleles from
two nearby genetic variants commonly occur together in a non-random,
linked fashion. This linked mode of inheritance results from genetic
variants in close proximity being less likely to be separated by a
recombination event and thus alleles of the variants are more commonly
inherited together than expected. Alleles of variants in linkage
disequilibrium are correlated; with the degree of correlation generally
greater in magnitude the closer the variants are in physical distance.
Measures of linkage disequilibrium include D prime (D’) and R squared
(R<sup>2</sup>).</p>
<p>A haplotype is a cluster of genetic variants that are inherited
together. Humans are diploid; having maternal and paternal copies of
each autosomal chromosome. Each chromosomal copy is organized into
segments of high linkage disequilibrium, called haplotype “blocks”. Due
to unique population histories and differences in variant allele
frequencies, haplotype structure tends to be population specific.
Although haplotypes are essential for calculating measures of linkage
disequilibrium, haplotypes are seldom directly observed. Statistical
chromosome phasing techniques are often necessary to infer individual
haplotypes.</p>
</div>
<div id="ldlink-data-sources" class="section level2">
<h2>LDlink Data Sources</h2>
<p><a href="https://www.ncbi.nlm.nih.gov/snp/">dbSNP</a> <a href="https://ftp.ncbi.nih.gov/snp/">(source: GRCh37 and GRCh38)</a> -
To investigate patterns of linkage disequilibrium, LDlink focuses on two
main classes of genetic variation: single nucleotide polymorphisms
(SNPs) and insertions/deletions (indels). Every module of LDlink
requires the entry of at least one variant as identified by a RefSNP
number (RS number) or genomic position (chr#:position). RS numbers are
unique labels assigned by dbSNP and are well-curated identifiers that
follow the format “rs” followed by a number. The current implementation
of LDlink references dbSNP and only accepts input for bi-allelic
variants.</p>
<p>1000 Genomes Project (source: <a href="http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/">GRCh37</a>,
<a href="http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/release/20190312_biallelic_SNV_and_INDEL/">GRCh38</a>,
and <a href="http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/working/20201028_3202_phased/">GRCh38
High Coverage</a>) - Publicly available reference haplotypes from the
1000 Genomes Project are used by LDlink to calculate population-specific
measures of linkage disequilibrium. Haplotypes are available for
continental populations (ex: European, African, and Admixed American)
and sub-populations (ex: Finnish, Gambian, and Peruvian). All LDlink
modules require the selection of at least one 1000 Genomes Project
sub-population, but several sub-populations can be selected
simultaneously. Available haplotypes vary by sub-population based on
sample size.</p>
<p><a href="https://genome.ucsc.edu/cgi-bin/hgTables">UCSC RefSeq</a>
(source: GRCh37 and GRCh38) - Publicly available gene transcripts from
the UCSC Table Browser are used by LDlink’s LDassoc (currently not
available in the <em>LDlinkR</em> package), LDmatrix, and LDproxy
modules to display genes within the genomic window of interest.</p>
<p><a href="https://regulomedb.org/">RegulomeDB</a> (source: GRCh37) -
Publicly available scores from RegulomeDB are used by LDlink’s LDassoc
(currently not available in the <em>LDlinkR</em> package) and LDproxy
modules to rank available data types for a single coordinate. GRCh38
support is added via <a href="https://genome.ucsc.edu/cgi-bin/hgLiftOver">liftOver</a>.</p>
<p><a href="https://mathgen.stats.ox.ac.uk/impute/1000GP_Phase3.html">Genetic
Map</a> (source: GRCh37) - Publicly available combined recombination
rates (cM/Mb) from the 1000 Genomes Project are used by LDlink’s LDassoc
(currently not available in the <em>LDlinkR</em> package) and LDproxy
modules to show recombination at specific coordinates. GRCh38 support is
added via <a href="https://genome.ucsc.edu/cgi-bin/hgLiftOver">liftOver</a>.</p>
<p><a href="https://www.gtexportal.org/home/">GTEx Portal</a> (source:
GRCh38) - Publicly available single-tissue cis-QTL data from the GTEx
Portal is used by LDlink’s LDexpress module to show significant
variant-gene associations in multiple tissue types. GRCh37 support is
added via GTEx lookup table.</p>
<p><a href="https://www.ebi.ac.uk/gwas/home">GWAS Catalog</a> (source:
GRCh38) - Publicly available NHGRI-EBI Catalog of human genome-wide
association studies from GWAS Catalog is used by LDlink’s LDtrait module
to search if variants have previously been associated with a trait or
disease. GRCh37 support is added via <a href="https://www.ncbi.nlm.nih.gov/snp/">dbSNP</a>.</p>
</div>
<div id="calculations" class="section level2">
<h2>Calculations</h2>
<p>LDlink modules report the following measures of linkage
disequilibrium: D prime, R squared, and goodness-of-fit statistics.
Below is a brief description of each measure.</p>
<p><strong>D prime (D’)</strong> - an indicator of allelic segregation
for two genetic variants. D’ values range from 0 to 1 with higher values
indicating tight linkage of alleles. A D’ value of 0 indicates no
linkage of alleles. A D’ value of 1 indicates at least one expected
haplotype combination is not observed.</p>
<p><strong>R squared (R<sup>2</sup>)</strong> - a measure of correlation
of alleles for two genetic variants. R<sup>2</sup> values range from 0
to 1 with higher values indicating a higher degree of correlation. An
R<sup>2</sup> value of 0 indicates alleles are independent, whereas an
R<sup>2</sup> value of 1 indicates an allele of one variant perfectly
predicts an allele of another variant. R<sup>2</sup> is sensitive to
allele frequency.</p>
<p><strong>Goodness of Fit (X<sup>2</sup> and p-value)</strong> -
statistical test testing whether observed haplotype counts follow
frequencies expected from variant allele frequencies. High chi-square
statistics and low p-values are evidence that haplotype counts deviate
from expected values and suggest linkage disequilibrium may be
present.</p>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<ul>
<li>The release version of <em>LDlinkR</em> can be installed from
CRAN:</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;LDlinkR&quot;</span>)</span></code></pre></div>
<ul>
<li>The development version of the <em>LDlinkR</em> package can be
installed from the GitHub repository by using the <em>remotes</em>
package:</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;remotes&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;CBIIT/LDlinkR&quot;</span>)</span></code></pre></div>
<p><em>LDlinkR</em> depends on the following packages:</p>
<ul>
<li><em>utils</em>, version 3.4.2 or later</li>
<li><em>httr</em>, version 1.4.0 or later</li>
</ul>
<p>Following installation, attach the <em>LDlinkR</em> package with:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(LDlinkR)</span></code></pre></div>
</div>
<div id="personal-access-token---required" class="section level2">
<h2>Personal Access Token - Required</h2>
<p>In order to access the LDlink API via <em>LDlinkR</em>, we use a
personal access token. This is a common convention followed by many APIs
and emulates the more familiar HTTPS username/password or SSH keys.</p>
<p>You will need to:</p>
<ul>
<li>Make a one-time request for your personal access token from a web
browser at <a href="https://ldlink.nih.gov/?tab=apiaccess" class="uri">https://ldlink.nih.gov/?tab=apiaccess</a>.</li>
<li>Once registered, your personal access token will be emailed to you.
It is a string of 12 random letters and numbers.</li>
<li>Provide your token as an argument when using <em>LDlinkR</em>. See
example below:</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDhap</span>(<span class="at">snps =</span> <span class="fu">c</span>(<span class="st">&quot;rs3&quot;</span>, <span class="st">&quot;rs4&quot;</span>, <span class="st">&quot;rs148890987&quot;</span>), </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">pop =</span> <span class="st">&quot;YRI&quot;</span>, </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">token =</span> <span class="st">&quot;YourTokenHere123&quot;</span>)</span></code></pre></div>
<p><strong>Optional:</strong> <br> However, the best security practice
is to store your personal access token as an environment variable where
<em>LDlinkR</em> can find it and use it on your behalf but where it will
not be accidentally shared with the public. <strong>Note:</strong>
Modifying R startup files (such as the <code>.Renviron</code>) is for
the advanced R user only. Modification of these files in the wrong way
could cause problems. Please proceed cautiously. Step-by-step
instructions follow:</p>
<p>After retrieving your personal access token from your email, put your
token in your <code>.Renviron</code> file. <code>.Renviron</code> is a
hidden file that lives in your home directory. The easiest way to both
find and edit the <code>.Renviron</code> file is with a function from
the <em>usethis</em> package. From the R console, do:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>usethis<span class="sc">::</span><span class="fu">edit_r_environ</span>()</span></code></pre></div>
<p>Your <code>.Renviron</code> file should open in your editor. Add a
line that looks like this:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>LDLINK_TOKEN<span class="ot">=</span>YourTokenHere123</span></code></pre></div>
<p><strong>Important</strong>, ensure you put a line break at the end by
hitting the <em>enter/return</em> key.</p>
<p>Save and close the <code>.Renviron</code> file. Restart R, as
environment variables are only loaded from <code>.Renviron</code> at the
start of a new R session. Now, check to see that your token is available
by entering:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;YourTokenHere123&quot;</code></pre>
<p>You should see your personal access token print to the screen, as
shown above. Now, <em>LDlinkR</em> function calls that use</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>)</span></code></pre></div>
<p>for the <code>token</code> argument in <em>LDlinkR</em> function
calls will use your personal access token in a private and secure way.
This method will be used in the extended examples that follow.</p>
<hr />
</div>
<div id="functions-and-examples" class="section level2">
<h2>Functions and Examples</h2>
<div id="ldexpress" class="section level3">
<h3>LDexpress</h3>
<div id="function" class="section level4">
<h4>Function</h4>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDexpress</span>(snps, </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">pop =</span> <span class="st">&quot;CEU&quot;</span>, </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>          <span class="at">tissue =</span> <span class="st">&quot;ALL&quot;</span>, </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">r2d =</span> <span class="st">&quot;r2&quot;</span>, </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">r2d_threshold =</span> <span class="fl">0.1</span>, </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">p_threshold =</span> <span class="fl">0.1</span>, </span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">win_size =</span> <span class="dv">500000</span>,</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">genome_build =</span> <span class="st">&quot;grch37&quot;</span>,</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">token =</span> <span class="cn">NULL</span>, </span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>          <span class="at">file =</span> <span class="cn">FALSE</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>         )</span></code></pre></div>
<p>Search if a list of genomic variants (or variants in LD with those
variants) is associated with gene expression in tissues of interest.
Quantitative trait loci data is downloaded from the <a href="https://gtexportal.org/home/">GTEx Portal</a>.</p>
</div>
<div id="arguments" class="section level4">
<h4>Arguments</h4>
<ul>
<li><code>snps</code>, between 1 - 10 variants, using an rsID or
chromosome coordinate (e.g. “chr7:24966446”)</li>
<li><code>pop</code>, a 1000 Genomes Project population, (e.g. YRI or
CEU), multiple allowed, default = “CEU”. See the <em>list_pop</em>
function in the utilities section below for available human populations
and their abbreviation codes.</li>
<li><code>tissue</code>, select from 1 - 54 non-diseased tissue sites
collected for the GTEx project, multiple allowed. Acceptable user input
is taken either from “tissue_name_ldexpress” or
“tissue_abbrev_ldexpress” (tissue abbreviation) code listed in available
GTEx tissue sites using the <em>list_gtex_tissues</em> function
(e.g. “ADI_SUB” for Adipose Subcutaneous). Input is case sensitive.
Default = “ALL” for all available tissue types.</li>
<li><code>r2d</code>, Select either “r2” for LD R<sup>2</sup>
(R-squared) or “d” for LD D’, default = “r2”.</li>
<li><code>r2d_threshold</code>, R-squared or D’ (depends on ‘r2d’ user
input parameter) threshold for LD filtering. Any variants within -/+ of
the specified genomic window and R<sup>2</sup> or D’ less than the
threshold will be removed. Value needs to be in the range 0 to 1.
Default value is 0.1.</li>
<li><code>p_threshold</code>, define the eQTL significance threshold
used for returning query results. Default value is 0.1 which returns all
GTEx eQTL associations with P-value less than 0.1.</li>
<li><code>win_size</code>, set genomic base pair window size for LD
calculation. Specify a value greater than or equal to zero and less than
or equal to 1000000 basepairs (bp). Default value is -/+ 500000 bp.</li>
<li><code>genome_build</code> Choose between one of three
options…<code>grch37</code> for genome build GRCh37 (hg19),
<code>grch38</code> for GRCh38 (hg38), or
<code>grch38_high_coverage</code> for GRCh38 High Coverage (hg38) 1000
Genome Project data sets. Default is GRCh37 (hg19).</li>
<li><code>token</code>, LDlink provided user access token is required,
default = NULL, register for a free token on the <a href="https://ldlink.nih.gov/?tab=apiaccess">LDlink web site.</a></li>
<li><code>file</code>, optional character string naming a path and file
for saving results. If file = FALSE, no file will be generated, default
= FALSE.</li>
</ul>
</div>
<div id="usage-single-query-variant-multiple-populations-multiple-tissue-types-using-tissue-abbreviation" class="section level4">
<h4>Usage: Single query variant, multiple populations, multiple tissue
types using tissue abbreviation</h4>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>my_output <span class="ot">&lt;-</span> <span class="fu">LDexpress</span>(<span class="at">snps =</span> <span class="st">&quot;rs4&quot;</span>,</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">pop =</span> <span class="fu">c</span>(<span class="st">&quot;YRI&quot;</span>, <span class="st">&quot;CEU&quot;</span>),</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">tissue =</span>  <span class="fu">c</span>(<span class="st">&quot;ADI_SUB&quot;</span>, <span class="st">&quot;ADI_VIS_OME&quot;</span>),</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">win_size =</span> <span class="st">&quot;500000&quot;</span>,</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>                      )</span></code></pre></div>
<p>In the above example, output is a data frame stored in the variable
<code>my_output</code>. See below.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(my_output)</span></code></pre></div>
<pre><code>##   Query      RS_ID Position_grch37                R2                D&#39;  Gene_Symbol        Gencode_ID
## 1   rs4 rs10637519  chr13:32430479 0.174249321651574 0.965976331360947 RP1-257C22.2 ENSG00000279314.1
## 2   rs4 rs10637519  chr13:32430479 0.174249321651574 0.965976331360947 RP1-257C22.2 ENSG00000279314.1
## 3   rs4   rs473641  chr13:32431244 0.174249321651574 0.965976331360947 RP1-257C22.2 ENSG00000279314.1
## 4   rs4   rs473641  chr13:32431244 0.174249321651574 0.965976331360947 RP1-257C22.2 ENSG00000279314.1
## 5   rs4   rs671746  chr13:32431263 0.174249321651574 0.965976331360947 RP1-257C22.2 ENSG00000279314.1
## 6   rs4   rs671746  chr13:32431263 0.174249321651574 0.965976331360947 RP1-257C22.2 ENSG00000279314.1
##                         Tissue Non_effect_Allele_Freq Effect_Allele_Freq Effect_Size     P_value
## 1 Adipose - Visceral (Omentum)                G=0.565          GTC=0.435    0.207161  1.0227e-05
## 2       Adipose - Subcutaneous                G=0.565          GTC=0.435    0.225642  2.2578e-07
## 3 Adipose - Visceral (Omentum)                A=0.565            G=0.435    0.207161  1.0227e-05
## 4       Adipose - Subcutaneous                A=0.565            G=0.435    0.225642  2.2578e-07
## 5 Adipose - Visceral (Omentum)                C=0.565            T=0.435    0.207161  1.0227e-05
## 6       Adipose - Subcutaneous                C=0.565            T=0.435    0.226558 1.93289e-07</code></pre>
</div>
<div id="usage-multiple-query-variants-single-population-a-tissue-type-using-the-full-ldexpress-tissue-name-no-spaces-and-genome-build-grch38." class="section level4">
<h4>Usage: Multiple query variants, single population, a tissue type
using the full LDexpress tissue name, no spaces, and genome build
GRCh38.</h4>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>my_output <span class="ot">&lt;-</span> <span class="fu">LDexpress</span>(<span class="at">snps =</span> <span class="fu">c</span>(<span class="st">&quot;rs345&quot;</span>, <span class="st">&quot;rs456&quot;</span>),</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">pop =</span> <span class="st">&quot;YRI&quot;</span>,</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">tissue =</span>  <span class="st">&quot;Adipose_Visceral_Omentum&quot;</span>,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">genome_build =</span> <span class="st">&quot;grch38&quot;</span>,</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>                      )</span></code></pre></div>
<p>In the above example, output is a data frame stored in the variable
<code>my_output</code>. See below.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(my_output)</span></code></pre></div>
<pre><code>##   Query      RS_ID Position_grch38                R2  D&#39; Gene_Symbol        Gencode_ID
## 1 rs345 rs12877069  chr13:32430415 0.222088835534214  1 RP1-257C22.2 ENSG00000279314.1
## 2 rs345 rs10637519  chr13:32430479  0.10989010989011  1 RP1-257C22.2 ENSG00000279314.1
## 3 rs345   rs473641  chr13:32431244  0.10989010989011  1 RP1-257C22.2 ENSG00000279314.1
## 4 rs345   rs671746  chr13:32431263  0.10989010989011  1 RP1-257C22.2 ENSG00000279314.1
## 5 rs345  rs9315146  chr13:32432193 0.222088835534214  1 RP1-257C22.2 ENSG00000279314.1
## 6 rs345   rs657190  chr13:32432232 0.107871720116618  1 RP1-257C22.2 ENSG00000279314.1
                           Tissue Non_effect_Allele_Freq Effect_Allele_Freq Effect_Size     P_value
## 1 Adipose - Visceral (Omentum)                C=0.685            T=0.315    0.355769 6.11598e-05
## 2 Adipose - Visceral (Omentum)                G=0.519          GTC=0.481    0.207161  1.0227e-05
## 3 Adipose - Visceral (Omentum)                A=0.519            G=0.481    0.207161  1.0227e-05
## 4 Adipose - Visceral (Omentum)                C=0.519            T=0.481    0.207161  1.0227e-05
## 5 Adipose - Visceral (Omentum)                A=0.685            G=0.315    0.276884 2.20517e-08
## 6 Adipose - Visceral (Omentum)                T=0.514            C=0.486    0.207916 9.95318e-06</code></pre>
<hr />
</div>
</div>
<div id="ldhap" class="section level3">
<h3>LDhap</h3>
<div id="function-1" class="section level4">
<h4>Function</h4>
<pre><code>LDhap(snps, 
      pop = &quot;CEU&quot;, 
      token = NULL, 
      file = FALSE, 
      table_type = &quot;haplotype&quot;,
      genome_build = &quot;grch37&quot;
     )</code></pre>
<p>Calculates population specific haplotype frequencies of all
haplotypes observed for a list of query variants. Input is a list of
variant RS numbers (concatenated list) and a population group.</p>
</div>
<div id="arguments-1" class="section level4">
<h4>Arguments</h4>
<ul>
<li><code>snps</code>, a list of between 1 - 30 variants, using an rsID
or chromosome coordinate (e.g. “chr7:24966446”)</li>
<li><code>pop</code>, a 1000 Genomes Project population, uses three
letter population code, (e.g. YRI or CEU), multiple allowed, default =
“CEU”</li>
<li><code>token</code>, LDlink provided user access token is required,
default = NULL, register for a free token on the <a href="https://ldlink.nih.gov/?tab=apiaccess">LDlink web site.</a></li>
<li><code>file</code>, optional character string naming a path and file
for saving results. If file = FALSE, no file will be generated, default
= FALSE</li>
<li><code>table_type</code>, Choose from one of four options available
to determine output format type…<code>haplotype</code>,
<code>variant</code>, <code>both</code> and <code>merged</code>. Default
= “haplotype”.</li>
<li><code>genome_build</code> Choose between one of three
options…<code>grch37</code> for genome build GRCh37 (hg19),
<code>grch38</code> for GRCh38 (hg38), or
<code>grch38_high_coverage</code> for GRCh38 High Coverage (hg38) 1000
Genome Project data sets. Default is GRCh37 (hg19).</li>
</ul>
</div>
<div id="usage-multiple-query-variants-single-population-and-genome-build-grch38-high-coverage" class="section level4">
<h4>Usage: Multiple query variants, single population, and genome build
GRCh38 High Coverage</h4>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDhap</span>(<span class="at">snps =</span> <span class="fu">c</span>(<span class="st">&quot;rs3&quot;</span>, <span class="st">&quot;rs4&quot;</span>, <span class="st">&quot;rs148890987&quot;</span>), </span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">pop =</span> <span class="st">&quot;CEU&quot;</span>, </span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>),</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">genome_build =</span> <span class="st">&quot;grch38_high_coverage&quot;</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>     )</span></code></pre></div>
<pre><code>##   rs3 rs4 rs148890987 Count Frequency
## 1   C   A           C   183    0.9242
## 2   T   G           C    15    0.0758</code></pre>
<p><br></p>
</div>
<div id="usage-multiple-query-variants-multiple-populations-using-default-genome-build-grch37-hg19" class="section level4">
<h4>Usage: Multiple query variants, multiple populations using default
genome build GRCh37 (hg19)</h4>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDhap</span>(<span class="at">snps =</span> <span class="fu">c</span>(<span class="st">&quot;rs3&quot;</span>, <span class="st">&quot;rs4&quot;</span>, <span class="st">&quot;rs148890987&quot;</span>),</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">pop =</span> <span class="fu">c</span>(<span class="st">&quot;YRI&quot;</span>, <span class="st">&quot;CEU&quot;</span>),</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>     )</span></code></pre></div>
<pre><code>##   rs3 rs4 rs148890987 Count Frequency
## 1   C   A           C   355    0.8575
## 2   T   G           C    41     0.099
## 3   T   G           T    11    0.0266
## 4   C   A           T     7    0.0169</code></pre>
<p>Output is a table of alleles, haplotype count and haplotype
frequencies.</p>
<p><br></p>
</div>
<div id="usage-multiple-query-variants-single-population-merged-output-format-type." class="section level4">
<h4>Usage: Multiple query variants, single population, ‘merged’ output
format type.</h4>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDhap</span>(<span class="at">snps =</span> <span class="fu">c</span>(<span class="st">&quot;rs660670&quot;</span>, <span class="st">&quot;rs556780&quot;</span>, <span class="st">&quot;rs355&quot;</span>, <span class="st">&quot;rs356&quot;</span>, <span class="st">&quot;rs542746&quot;</span>),</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">pop =</span> <span class="st">&quot;CEU&quot;</span>,</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>),</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">table_type =</span> <span class="st">&quot;merged&quot;</span>,</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">genome_build =</span> <span class="st">&quot;grch38&quot;</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>     )</span></code></pre></div>
<pre><code>##   RS_Number Position_grch38    Allele_Frequency Haplotypes       
## 1  rs660670  chr13:31863887    A=0.924, G=0.076          A      G
## 2  rs556780  chr13:31863023    G=0.924, A=0.076          G      A
## 3     rs355  chr13:31883842    A=0.924, G=0.076          A      G
## 4     rs356  chr13:31884663    T=0.924, A=0.076          T      A
## 5  rs542746  chr13:31860055    G=0.924, A=0.076          G      A
## 6                               Haplotype_Count        183     15
## 7                           Haplotype_Frequency     0.9242 0.0758</code></pre>
<p>Output is a table with query variants, genomic position GRCH38
(hg38), etc.</p>
<p><br></p>
</div>
<div id="usage-multiple-query-variants-single-population-both-output-format-type." class="section level4">
<h4>Usage: Multiple query variants, single population, ‘both’ output
format type.</h4>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDhap</span>(<span class="at">snps =</span> <span class="fu">c</span>(<span class="st">&quot;rs660670&quot;</span>, <span class="st">&quot;rs556780&quot;</span>, <span class="st">&quot;rs355&quot;</span>, <span class="st">&quot;rs356&quot;</span>, <span class="st">&quot;rs542746&quot;</span>),</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">pop =</span> <span class="st">&quot;CEU&quot;</span>,</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>),</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">table_type =</span> <span class="st">&quot;both&quot;</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>     )</span></code></pre></div>
<pre><code>## [[1]]
##   RS_Number Position_grch37 Allele_Frequency
## 1  rs660670  chr13:32438024 A=0.924, G=0.076
## 2  rs556780  chr13:32437160 G=0.924, A=0.076
## 3     rs355  chr13:32457979 A=0.924, G=0.076
## 4     rs356  chr13:32458800 T=0.924, A=0.076
## 5  rs542746  chr13:32434192 G=0.924, A=0.076
## 
## [[2]]
##   rs660670 rs556780 rs355 rs356 rs542746 Count Frequency
## 1        A        G     A     T        G   183    0.9242
## 2        G        A     G     A        A    15    0.0758</code></pre>
<p>Output is a list that contains both the ‘variant’ and ‘haplotype’
output format types.</p>
<hr />
</div>
</div>
<div id="ldmatrix" class="section level3">
<h3>LDmatrix</h3>
<div id="function-2" class="section level4">
<h4>Function</h4>
<pre><code>LDmatrix(snps, 
         pop = &quot;CEU&quot;, 
         r2d = &quot;r2&quot;, 
         token = NULL, 
         file = FALSE,
         genome_build = &quot;grch37&quot;
        )</code></pre>
<p>Generates a data frame of pairwise linkage disequilibrium statistics.
Input is a list of between 2 to 1000 variants. Desired output can be
based on estimates of R<sup>2</sup> or D’.</p>
</div>
<div id="arguments-2" class="section level4">
<h4>Arguments</h4>
<ul>
<li><code>snps</code>, list of between 2 - 1,000 variants, using an rsID
or chromosome coordinate (GRCh37/hg19) (e.g. “chr7:24966446”)</li>
<li><code>pop</code>, a 1000 Genomes Project population, uses three
letter population code, (e.g. YRI or CEU), multiple allowed, default =
“CEU”</li>
<li><code>r2d</code>, use either “r2” for pairwise R<sup>2</sup>
statistics or “d” for pairwise D’ statistics</li>
<li><code>token</code>, LDlink provided user access token is required,
default = NULL, register for a free token on the <a href="https://ldlink.nih.gov/?tab=apiaccess">LDlink web site.</a></li>
<li><code>file</code>, optional character string naming a path and file
for saving results. If file = FALSE, no file will be generated, default
= FALSE</li>
<li><code>genome_build</code> Choose between one of three
options…<code>grch37</code> for genome build GRCh37 (hg19),
<code>grch38</code> for GRCh38 (hg38), or
<code>grch38_high_coverage</code> for GRCh38 High Coverage (hg38) 1000
Genome Project data sets. Default is GRCh37 (hg19).</li>
</ul>
</div>
<div id="usage-multiple-query-variants-single-population-r2-and-genome-build-grch38-hg38." class="section level4">
<h4>Usage: Multiple query variants, single population, R<sup>2</sup> and
genome build GRCh38 (hg38).</h4>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDmatrix</span>(<span class="at">snps =</span> <span class="fu">c</span>(<span class="st">&quot;rs496202&quot;</span>, <span class="st">&quot;rs11147477&quot;</span>, <span class="st">&quot;rs201578600&quot;</span>), </span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">pop =</span> <span class="st">&quot;YRI&quot;</span>, </span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">r2d =</span> <span class="st">&quot;r2&quot;</span>, </span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>),</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">genome_build =</span> <span class="st">&quot;grch38&quot;</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>        )</span></code></pre></div>
<pre><code>##     RS_number rs496202 rs11147477 rs201578600
## 1    rs496202    1.000      0.503       0.659
## 2  rs11147477    0.503      1.000       0.786
## 3 rs201578600    0.659      0.786       1.000</code></pre>
<p><br></p>
</div>
<div id="usage-multiple-query-variants-mixed-use-of-rsids-genomic-coordinates-multiple-populations-d" class="section level4">
<h4>Usage: Multiple query variants (mixed use of rsID’s &amp; genomic
coordinates), multiple populations, D’</h4>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDmatrix</span>(<span class="at">snps =</span> <span class="fu">c</span>(<span class="st">&quot;chr13:32444611&quot;</span>, <span class="st">&quot;rs11147477&quot;</span>, <span class="st">&quot;rs201578600&quot;</span>), </span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">pop =</span> <span class="fu">c</span>(<span class="st">&quot;YRI&quot;</span>, <span class="st">&quot;CEU&quot;</span>), </span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">r2d =</span> <span class="st">&quot;d&quot;</span>, </span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>        )</span></code></pre></div>
<pre><code>##     RS_number rs496202 rs11147477 rs201578600
## 1    rs496202    1.000      0.738       0.973
## 2  rs11147477    0.738      1.000       0.971
## 3 rs201578600    0.973      0.971       1.000</code></pre>
<p><br></p>
</div>
<div id="usage-multiple-query-variants-read-from-text-file-multiple-populations-d" class="section level4">
<h4>Usage: Multiple query variants read from text file, multiple
populations, D’</h4>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>my_variants <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;variant_list.txt&quot;</span>)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>my_variants</span></code></pre></div>
<pre><code>##           V1
## 1      rs456
## 2      rs114
## 3      rs127
## 4  rs7805287
## 5 rs60676332
## 6 rs10239961</code></pre>
<p>Then, call <em>LDmatrix</em> with:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDmatrix</span>(<span class="at">snps =</span> my_variants[,<span class="dv">1</span>], </span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">pop =</span> <span class="fu">c</span>(<span class="st">&quot;YRI&quot;</span>, <span class="st">&quot;CEU&quot;</span>), <span class="at">r2d =</span> <span class="st">&quot;d&quot;</span>, </span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>        )</span></code></pre></div>
<pre><code>##   RS_number rs456 rs114 rs127 rs7805287 rs60676332 rs10239961
## 1      rs456 1.000 0.963 0.929     0.789      0.151      1.000
## 2      rs114 0.963 1.000 0.886     0.710      0.148      0.459
## 3      rs127 0.929 0.886 1.000     0.818      0.180      0.912
## 4  rs7805287 0.789 0.710 0.818     1.000      0.094      0.464
## 5 rs60676332 0.151 0.148 0.180     0.094      1.000      0.363
## 6 rs10239961 1.000 0.459 0.912     0.464      0.363      1.000</code></pre>
<p>Output is a table with rows and columns equal to the number of query
variants and pairwise linkage disequilibrium statistics.</p>
<hr />
</div>
</div>
<div id="ldpair" class="section level3">
<h3>LDpair</h3>
<div id="function-3" class="section level4">
<h4>Function</h4>
<pre><code>LDpair(var1, 
       var2, 
       pop = &quot;CEU&quot;, 
       token = NULL, 
       output = &quot;table&quot;, 
       file = FALSE,
       genome_build = &quot;grch37&quot;
      )</code></pre>
<p>Investigates potentially correlated alleles for a pair of variants.
Input is two query variants and a 1000 Genomes Project reference
population(s) of interest.</p>
</div>
<div id="arguments-3" class="section level4">
<h4>Arguments</h4>
<ul>
<li><code>var1</code>, the first RS number (rsID) or genomic coordinate
(GRCh37/hg19) (e.g. “chr7:24966446”), must match a bi-allelic
variant</li>
<li><code>var2</code>, the second RS number or genomic coordinate, as
above, must match a bi-allelic variant</li>
<li><code>pop</code>, a 1000 Genomes Project reference population, uses
three letter population code, (e.g. YRI or CEU), multiple allowed,
default = “CEU”</li>
<li><code>token</code>, LDlink provided user access token is required,
default = NULL, register for a free token on the <a href="https://ldlink.nih.gov/?tab=apiaccess">LDlink web site.</a></li>
<li><code>output</code>, two output format options are available,
“text”, which displays a two-by-two matrix displaying haplotype counts
and allele frequencies along with other statistics, or “table”, which
displays the same data in rows and columns, default = “table”</li>
<li><code>file</code>, optional character string naming a path and file
for saving results. If file = FALSE, no file will be generated, default
= FALSE</li>
<li><code>genome_build</code> Choose between one of three
options…<code>grch37</code> for genome build GRCh37 (hg19),
<code>grch38</code> for GRCh38 (hg38), or
<code>grch38_high_coverage</code> for GRCh38 High Coverage (hg38) 1000
Genome Project data sets. Default is GRCh37 (hg19).</li>
</ul>
</div>
<div id="usage-with-output-argument-set-to-text-and-genome-build-grch38-hg38" class="section level4">
<h4>Usage: With <code>output</code> argument set to “text” and genome
build GRCh38 (hg38)</h4>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDpair</span>(<span class="at">var1 =</span> <span class="st">&quot;rs496202&quot;</span>, </span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">var2 =</span> <span class="st">&quot;rs11147477&quot;</span>, </span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">pop =</span> <span class="st">&quot;YRI&quot;</span>, </span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>), </span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">output =</span> <span class="st">&quot;text&quot;</span>,</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">genome_build =</span> <span class="st">&quot;grch38&quot;</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>      )</span></code></pre></div>
<pre><code>## Query SNPs:
## rs496202 (chr13:32444611)
## rs11147477 (chr13:32509120)
## 
## YRI Haplotypes:
##                rs11147477
##                C       T
##              -----------------
##            C | 11    | 26    | 37    (0.171)
## rs496202     -----------------
##            G | 173   | 6     | 179   (0.829)
##              -----------------
##                184     32      216
##               (0.852) (0.148)
## 
##           G_C: 173 (0.801)
##           C_T: 26 (0.12)
##           C_C: 11 (0.051)
##           G_T: 6 (0.028)
## 
##           D&#39;: 0.7737
##           R2: 0.5037
##       Chi-sq: 108.8005
##      p-value: &lt;0.0001
## 
## rs496202(C) allele is correlated with rs11147477(T) allele
## rs496202(G) allele is correlated with rs11147477(C) allele</code></pre>
<p><br></p>
</div>
<div id="usage-with-no-output-argument-option-specified-using-default-table." class="section level4">
<h4>Usage: With no <code>output</code> argument option specified, using
default “table”.</h4>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDpair</span>(<span class="at">var1 =</span> <span class="st">&quot;rs496202&quot;</span>, </span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">var2 =</span> <span class="st">&quot;rs11147477&quot;</span>, </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">pop =</span> <span class="st">&quot;YRI&quot;</span>, </span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>       <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>),</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">genome_build =</span> <span class="st">&quot;grch38&quot;</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>      )</span></code></pre></div>
<pre><code>##       var1       var2 pops       var1_pos       var2_pos var1_a1 var1_a2 var1_a1_freq var1_a2_freq var2_a1
## 1 rs496202 rs11147477  YRI chr13:31870474 chr13:31934983       C       G        0.173        0.827       C
##   var2_a2 var2_a1_freq var2_a2_freq d_prime    r2   chisq p_val
## 1       T         0.85         0.15  0.7733 0.503 107.638 1e-04
##                                           corr_alleles
## 1 rs496202(C)-rs11147477(T), rs496202(G)-rs11147477(C)</code></pre>
<p>Output of the <code>output</code> argument “text” option is a
two-by-two contingency table displaying haplotype counts and allele
frequencies of the two query variants. Also displayed are calculated
metrics of linkage disequilibrium including: D prime (D’), R square
(R<sup>2</sup>), and goodness-of-fit (Chi-square and p-value).
Goodness-of-fit tests for deviations of expected haplotype frequencies
based on allele frequencies. Correlated alleles are reported if linkage
disequilibrium is present (R<sup>2</sup> &gt; 0.1). If linkage
equilibrium, no alleles are reported.</p>
<p>Output from the <code>output</code> argument “table” option converts
the data from the two-by-two contingency table into a data frame.</p>
<hr />
</div>
</div>
<div id="ldpop" class="section level3">
<h3>LDpop</h3>
<div id="function-4" class="section level4">
<h4>Function</h4>
<pre><code>LDpop(var1, 
      var2, 
      pop = &quot;CEU&quot;, 
      r2d = &quot;r2&quot;, 
      token = NULL, 
      file = FALSE,
      genome_build = &quot;grch37&quot;
     )</code></pre>
<p>Investigates allele frequencies and linkage disequilibrium patterns
across 1000G populations.</p>
</div>
<div id="arguments-4" class="section level4">
<h4>Arguments</h4>
<ul>
<li><code>var1</code>, the first RS number (rsID) or genomic coordinate
(GRCh37/hg19) (e.g. “chr7:24966446”), must match a bi-allelic
variant</li>
<li><code>var2</code>, the second RS number or genomic coordinate, as
above, must match a bi-allelic variant</li>
<li><code>pop</code>, a 1000 Genomes Project reference population, uses
three letter population code, (e.g. YRI or CEU), multiple allowed,
default = “CEU”</li>
<li><code>r2d</code>, use “r2” if desired output is based on estimated
R<sup>2</sup> or “d” if D’</li>
<li><code>token</code>, LDlink provided user access token is required,
default = NULL, register for a free token on the <a href="https://ldlink.nih.gov/?tab=apiaccess">LDlink web site.</a></li>
<li><code>file</code>, optional character string naming a path and file
for saving results. If file = FALSE, no file will be generated, default
= FALSE</li>
<li><code>genome_build</code> Choose between one of three
options…<code>grch37</code> for genome build GRCh37 (hg19),
<code>grch38</code> for GRCh38 (hg38), or
<code>grch38_high_coverage</code> for GRCh38 High Coverage (hg38) 1000
Genome Project data sets. Default is GRCh37 (hg19).</li>
</ul>
</div>
<div id="usage" class="section level4">
<h4>Usage</h4>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDpop</span>(<span class="at">var1 =</span> <span class="st">&quot;rs496202&quot;</span>, </span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">var2 =</span> <span class="st">&quot;rs11147477&quot;</span>, </span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">pop =</span> <span class="st">&quot;YRI&quot;</span>, </span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">r2d =</span> <span class="st">&quot;r2&quot;</span>, </span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>),</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">genome_build =</span> <span class="st">&quot;grch38_high_coverage&quot;</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>     )</span></code></pre></div>
<pre><code>##                  Population Abbrev   N rs496202_Allele_Freq rs11147477_Allele_Freq     R2     D&#39;    Chisq P
## 1 Yoruba in Ibadan, Nigeria    YRI 108 G: 82.87%, C: 17.13%   C: 85.19%, T: 14.81% 0.5037 0.7737 108.8005 0</code></pre>
<hr />
</div>
</div>
<div id="ldproxy" class="section level3">
<h3>LDproxy</h3>
<div id="function-5" class="section level4">
<h4>Function</h4>
<pre><code>LDproxy(snp, 
        pop = &quot;CEU&quot;, 
        r2d = &quot;r2&quot;, 
        token = NULL, 
        file = FALSE,
        genome_build = &quot;grch37&quot;
       )</code></pre>
<p>Explore proxy and putative functional variants for a single query
variant. Input is a single RS number and a population group. Depending
on the number of query populations, this function could take some time
to run.</p>
</div>
<div id="arguments-5" class="section level4">
<h4>Arguments</h4>
<ul>
<li><code>snp</code>, an RS number (rsID) or chromosome coordinate
(GRCh37/hg19) (e.g. “chr7:24966446”), one per query, RS number must
match a bi-allelic variant</li>
<li><code>pop</code>, a 1000 Genomes Project reference population, uses
three letter population code, (e.g. YRI or CEU), multiple allowed,
default = “CEU”</li>
<li><code>r2d</code>, use “r2” if desired output is based on estimated
R<sup>2</sup> or “d” if D’</li>
<li><code>token</code>, LDlink provided user access token is required,
default = NULL, register for a free token on the <a href="https://ldlink.nih.gov/?tab=apiaccess">LDlink web site.</a></li>
<li><code>file</code>, optional character string naming a path and file
for saving results. If file = FALSE, no file will be generated, default
= FALSE</li>
<li><code>genome_build</code> Choose between one of three
options…<code>grch37</code> for genome build GRCh37 (hg19),
<code>grch38</code> for GRCh38 (hg38), or
<code>grch38_high_coverage</code> for GRCh38 High Coverage (hg38) 1000
Genome Project data sets. Default is GRCh37 (hg19).</li>
</ul>
</div>
<div id="usage-single-reference-population-and-default-genome-build-grch37-hg19." class="section level4">
<h4>Usage: single reference population and default genome build GRCh37
(hg19).</h4>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>my_proxies <span class="ot">&lt;-</span> <span class="fu">LDproxy</span>(<span class="at">snp =</span> <span class="st">&quot;rs456&quot;</span>, </span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">pop =</span> <span class="st">&quot;YRI&quot;</span>, </span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">r2d =</span> <span class="st">&quot;r2&quot;</span>, </span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>)</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>                     )</span></code></pre></div>
<p>Output is a data frame stored in the variable <code>my_proxies</code>
with 2455 rows and 10 columns with data.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(my_proxies)</span></code></pre></div>
<pre><code>##    RS_Number         Coord Alleles    MAF Distance Dprime     R2
## 1      rs456 chr7:24962419   (G/C) 0.1944        0      1 1.0000
## 2      rs457 chr7:24962426   (T/C) 0.1944        7      1 1.0000
## 3 rs28475742 chr7:24964633   (G/T) 0.1944     2214      1 1.0000
## 4      rs123 chr7:24966446   (C/A) 0.1944     4027      1 1.0000
## 5      rs125 chr7:24959703   (C/T) 0.2037    -2716      1 0.9436
## 6      rs128 chr7:24958977   (C/T) 0.2037    -3442      1 0.9436
##   Correlated_Alleles RegulomeDB Function
## 1            G=G,C=C          4     &lt;NA&gt;
## 2            G=T,C=C         2b     &lt;NA&gt;
## 3            G=G,C=T          4     &lt;NA&gt;
## 4            G=C,C=A         1f     &lt;NA&gt;
## 5            G=C,C=T         3a     &lt;NA&gt;
## 6            G=C,C=T          7     &lt;NA&gt;</code></pre>
<p>Includes information on all variants -/+ 500 Kb of the query variant
with a pairwise R<sup>2</sup> value greater than 0.01.</p>
<hr />
</div>
</div>
<div id="ldproxy_batch" class="section level3">
<h3>LDproxy_batch</h3>
<div id="function-6" class="section level4">
<h4>Function</h4>
<pre><code>LDproxy_batch(snp, 
              pop = &quot;CEU&quot;, 
              r2d = &quot;r2&quot;, 
              token = NULL, 
              append = FALSE,
              genome_build = &quot;grch37&quot;
             )</code></pre>
<p>Query <em>LDproxy</em> using a list of query variants.
<em>LDproxy_batch</em> will make sequential queries, one query per
variant. Concurrent queries are not permitted by the LDlink API. Output
is saved as text file(s) to the current working directory. Depending on
the number of query variants and reference populations selected, this
function could time some time to run.</p>
</div>
<div id="arguments-6" class="section level4">
<h4>Arguments</h4>
<ul>
<li><code>snp</code>, a character string or data frame listing RS
numbers (rsID) or chromosome coordinates (GRCh37/hg19)
(e.g. “chr7:24966446”), one per line.</li>
<li><code>pop</code>, a 1000 Genomes Project reference population, uses
three letter population code, (e.g. YRI or CEU), multiple allowed,
default = “CEU”</li>
<li><code>r2d</code>, use “r2” if desired output is based on estimated
R<sup>2</sup> or “d” if D’</li>
<li><code>token</code>, LDlink provided user access token is required,
default = NULL, register for a free token on the <a href="https://ldlink.nih.gov/?tab=apiaccess">LDlink web site.</a></li>
<li><code>append</code>, a logical, if TRUE, output for each query
variant is appended to a single text file and saved to the current
working directory. If FALSE, output for each query variant is saved in
its own text file with the query variant as the filename. Default value
is FALSE.</li>
<li><code>genome_build</code> Choose between one of three
options…<code>grch37</code> for genome build GRCh37 (hg19),
<code>grch38</code> for GRCh38 (hg38), or
<code>grch38_high_coverage</code> for GRCh38 High Coverage (hg38) 1000
Genome Project data sets. Default is GRCh37 (hg19).</li>
</ul>
</div>
<div id="usage-multiple-variants-default-pop-and-r2d" class="section level4">
<h4>Usage: multiple variants, default <code>pop</code> and
<code>r2d</code></h4>
<p>The list of query variants passed to <em>LDproxy_batch</em> can be
stored as a character string.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDproxy_batch</span>(<span class="at">snp =</span> <span class="fu">c</span>(<span class="st">&quot;rs456&quot;</span>, <span class="st">&quot;rs114&quot;</span>, <span class="st">&quot;rs127&quot;</span>), </span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>)</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>             )</span></code></pre></div>
<p>Or, a longer list of variants can be read into a data frame from a
text file and passed into <em>LDproxy_batch</em>. The list should be in
a simple text file, one query variant per line. For example:</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>my_variants <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;variant_list.txt&quot;</span>)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>my_variants</span></code></pre></div>
<pre><code>##           V1
## 1      rs456
## 2      rs114
## 3      rs127
## 4  rs7805287
## 5 rs60676332
## 6 rs10239961</code></pre>
<p>Then, call <em>LDproxy_batch</em> with:</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDproxy_batch</span>(<span class="at">snp =</span> my_variants, </span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>)</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>             )</span></code></pre></div>
<p>Output not displayed. All output from <em>LDproxy_batch</em> is saved
to a text file(s) in the current working directory.</p>
<hr />
</div>
</div>
<div id="ldtrait" class="section level3">
<h3>LDtrait</h3>
<div id="function-7" class="section level4">
<h4>Function</h4>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDtrait</span>(snps,</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">pop =</span> <span class="st">&quot;CEU&quot;</span>,</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">r2d =</span> <span class="st">&quot;r2&quot;</span>,</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">r2d_threshold =</span> <span class="fl">0.1</span>,</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">win_size =</span> <span class="dv">500000</span>,</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">token =</span> <span class="cn">NULL</span>,</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">file =</span> <span class="cn">FALSE</span>,</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">genome_build =</span> <span class="st">&quot;grch37&quot;</span></span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>       )</span></code></pre></div>
<p>Search if a list of variants (or variants in LD with those variants)
have been previously associated with a trait or disease. Trait and
disease data is updated nightly from the <a href="https://www.ebi.ac.uk/gwas/docs/file-downloads">GWAS
Catalog</a>.</p>
</div>
<div id="arguments-7" class="section level4">
<h4>Arguments</h4>
<ul>
<li><code>snps</code>, between 1 - 50 variants, using an rsID or
chromosome coordinate (GRCh37)(e.g. “chr7:24966446”). All input variants
must match a bi-allelic variant.</li>
<li><code>pop</code>, a 1000 Genomes Project population, (e.g. YRI or
CEU), multiple allowed, default = “CEU”. See the <em>list_pop</em>
function in the utilities section below for available human populations
and their abbreviation codes.</li>
<li><code>r2d</code>, use “r2” to filter desired output from a threshold
based on estimated LD R2 (R squared) or “d” for LD D’ (D-prime), default
= “r2”.</li>
<li><code>r2d_threshold</code>, R-squared or D’ (depends on ‘r2d’ user
input parameter) threshold for LD filtering. Any variants within -/+ of
the specified genomic window and R<sup>2</sup> or D’ less than the
threshold will be removed. Value needs to be in the range 0 to 1.
Default value is 0.1.</li>
<li><code>win_size</code>, set genomic base pair window size for LD
calculation. Specify a value greater than or equal to zero and less than
or equal to 1000000 basepairs (bp). Default value is -/+ 500000 bp.</li>
<li><code>token</code>, LDlink provided user access token is required,
default = NULL, register for a free token on the <a href="https://ldlink.nih.gov/?tab=apiaccess">LDlink web site.</a>.</li>
<li><code>file</code>, optional character string naming a path and file
for saving results. If file = FALSE, no file will be generated, default
= FALSE.</li>
<li><code>genome_build</code> Choose between one of three
options…<code>grch37</code> for genome build GRCh37 (hg19),
<code>grch38</code> for GRCh38 (hg38), or
<code>grch38_high_coverage</code> for GRCh38 High Coverage (hg38) 1000
Genome Project data sets. Default is GRCh37 (hg19).</li>
</ul>
</div>
<div id="usage-single-query-variant-multiple-reference-populations-and-genome-build-grch38hg38" class="section level4">
<h4>Usage: Single query variant, multiple reference populations and
genome build GRCh38(hg38)</h4>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDtrait</span>(<span class="at">snps =</span> <span class="st">&quot;rs456&quot;</span>,</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">pop =</span> <span class="fu">c</span>(<span class="st">&quot;YRI&quot;</span>, <span class="st">&quot;CEU&quot;</span>),</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>),</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">genome_build =</span> <span class="st">&quot;grch38&quot;</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>       )</span></code></pre></div>
<p>The following is the output from the above function call.</p>
<pre><code>##   Query                      GWAS_Trait  RS_Number Position_GRCh38          Alleles
## 1 rs456 Highest math class taken (MTAG) rs10248878   chr7:24869118 C=0.175, T=0.825
## 2 rs456   Educational attainment (MTAG)      rs457   chr7:24922807 C=0.697, T=0.303
##                 R2                D&#39; Risk_Allele Effect_Size_95_CI    Beta_or_OR P_value
## 1 0.41175133337888 0.920247773906311      0.5967            0.0104 0.0071-0.0137   7e-10
## 2                1                 1      0.4495            0.0072 0.0047-0.0097   4e-08</code></pre>
</div>
<div id="usage-multiple-query-variants-multiple-reference-populations-and-win_size-set-to-750000-basepairs-bp-default-genome-build-grch37hg19." class="section level4">
<h4>Usage: Multiple query variants, multiple reference populations and
<em>win_size</em> set to 750000 basepairs (bp), default genome build
GRCh37(hg19).</h4>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LDtrait</span>(<span class="at">snps =</span> <span class="fu">c</span>(<span class="st">&quot;rs114&quot;</span>, <span class="st">&quot;rs496202&quot;</span>, <span class="st">&quot;rs345&quot;</span>),</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">pop =</span> <span class="fu">c</span>(<span class="st">&quot;YRI&quot;</span>, <span class="st">&quot;CHB&quot;</span>, <span class="st">&quot;CEU&quot;</span>),</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">win_size =</span> <span class="st">&quot;750000&quot;</span>,</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>)</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>       )</span></code></pre></div>
<p>Output of the above function is below.</p>
<pre><code>##      Query                                        GWAS_Trait  RS_Number
## 1    rs114                   Highest math class taken (MTAG) rs10248878
## 2    rs114                     Educational attainment (MTAG)      rs457
## 3 rs496202                                  Refractive error      rs353
## 4    rs345            DNA methylation variation (age effect)   rs203425
## 5    rs345 Facial morphology (factor 14, intercanthal width)   rs799522
##   Position_GRCh37          Alleles                R2                D&#39;
## 1   chr7:24908737 C=0.123, T=0.877 0.200231693692643 0.897255733792921
## 2   chr7:24962426 C=0.748, T=0.252  0.56312684849231 0.969967060647161
## 3  chr13:32454349 A=0.902, G=0.098                 1                 1
## 4  chr13:32468087 A=0.074, T=0.926 0.954994192799071                 1
## 5  chr13:32514028 C=0.769, T=0.231 0.236284178064096 0.918763102725367
##   Risk_Allele Effect_Size_95_CI    Beta_or_OR P_value
## 1      0.5967            0.0104 0.0071-0.0137   7e-10
## 2      0.4495            0.0072 0.0047-0.0097   4e-08
## 3        &lt;NA&gt;              &lt;NA&gt;          &lt;NA&gt;   1e-12
## 4          NR              &lt;NA&gt;          &lt;NA&gt;   2e-08
## 5      0.1263            0.2157     0.12-0.31   6e-06</code></pre>
<hr />
</div>
</div>
<div id="snpchip" class="section level3">
<h3>SNPchip</h3>
<div id="function-8" class="section level4">
<h4>Function</h4>
<pre><code>SNPchip(snps, 
        chip = &quot;ALL&quot;, 
        token = NULL, 
        file = FALSE,
        genome_build = &quot;grch37&quot;
       )</code></pre>
<p>Used to find commercial genotyping chip arrays for variants. Input is
a list of between 1 - 5000 variants (one per line) and desired
commercial chip arrays to search. Input variants do not need to be on
the same chromosome.</p>
</div>
<div id="arguments-8" class="section level4">
<h4>Arguments</h4>
<ul>
<li><code>snps</code>, between 1 - 5,000 variants, using an rsID or
chromosome coordinate (e.g. “chr7:24966446”)</li>
<li><code>chip</code>, chip or arrays, platform code(s) for a SNP chip
array, ALL_Illumina, ALL_Affy or ALL, default=ALL, use the
<code>list_chips</code> utility (see below) to lookup available
commercial SNP chip arrays and their codes.</li>
<li><code>token</code>, LDlink provided user access token is required,
default = NULL, register for a free token on the <a href="https://ldlink.nih.gov/?tab=apiaccess">LDlink web site.</a></li>
<li><code>file</code>, optional character string naming a path and file
for saving results. If file = FALSE, no file will be generated, default
= FALSE.</li>
<li><code>genome_build</code> Choose between one of three
options…<code>grch37</code> for genome build GRCh37 (hg19),
<code>grch38</code> for GRCh38 (hg38), or
<code>grch38_high_coverage</code> for GRCh38 High Coverage (hg38) 1000
Genome Project data sets. Default is GRCh37 (hg19).</li>
</ul>
</div>
<div id="usage-multiple-variants-search-all-available-chip-arrays" class="section level4">
<h4>Usage: Multiple variants, search “ALL” available chip arrays</h4>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SNPchip</span>(<span class="at">snps =</span> <span class="fu">c</span>(<span class="st">&quot;rs3&quot;</span>, <span class="st">&quot;rs4&quot;</span>, <span class="st">&quot;rs148890987&quot;</span>), </span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">chip =</span> <span class="st">&quot;ALL&quot;</span>, </span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>)</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>       )</span></code></pre></div>
<pre><code>## WARNING: The following RS number did not have any platforms found: rs148890987, rs3.

##     RS_Number Position_GRCh37 A_SNP5.0 A_CHB2 A_250S A_SNP6.0
## 1 rs148890987  chr13:32403784        0      0      0        0
## 2         rs3  chr13:32446842        0      0      0        0
## 3         rs4  chr13:32447222        1      1      1        1</code></pre>
<p><br></p>
</div>
<div id="usage-multiple-variants-search-two-affymetrix-arrays-and-genome-build-grch38-hg38" class="section level4">
<h4>Usage: Multiple variants, search two Affymetrix arrays, and genome
build GRCh38 (hg38)</h4>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SNPchip</span>(<span class="at">snps =</span> <span class="fu">c</span>(<span class="st">&quot;rs3&quot;</span>, <span class="st">&quot;rs4&quot;</span>, <span class="st">&quot;rs148890987&quot;</span>), </span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">chip =</span> <span class="fu">c</span>(<span class="st">&quot;A_SNP5.0&quot;</span>, <span class="st">&quot;A_CHB2&quot;</span>), </span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>),</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">genome_build =</span> <span class="st">&quot;grch38&quot;</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>       )</span></code></pre></div>
<pre><code>## WARNING: The following RS number did not have any platforms found: rs148890987, rs3.

##     RS_Number Position_GRCh38 A_SNP5.0 A_CHB2
## 1 rs148890987     13:31829647        0      0
## 2         rs3     13:31872705        0      0
## 3         rs4     13:31873085        1      1</code></pre>
<p><br></p>
</div>
<div id="usage-multiple-variants-search-all-available-affymetrix-arrays-using-all_affy" class="section level4">
<h4>Usage: Multiple variants, search all available Affymetrix arrays
using, “ALL_Affy”</h4>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SNPchip</span>(<span class="at">snps =</span> <span class="fu">c</span>(<span class="st">&quot;rs3&quot;</span>, <span class="st">&quot;rs4&quot;</span>, <span class="st">&quot;rs148890987&quot;</span>), </span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">chip =</span> <span class="st">&quot;ALL_Affy&quot;</span>, </span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>)</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>       )</span></code></pre></div>
<pre><code>## WARNING: The following RS number did not have any platforms found: rs148890987, rs3.

##     RS_Number Position_GRCh37 A_SNP5.0 A_CHB2 A_250S A_SNP6.0
## 1 rs148890987  chr13:32403784        0      0      0        0
## 2         rs3  chr13:32446842        0      0      0        0
## 3         rs4  chr13:32447222        1      1      1        1</code></pre>
<p>Output is a data frame of query variant rows (RS number), genomic
coordinate (GRCh37) and genotyping chip array columns. The presence of a
“1” designates the variant is present on the respective commercial
genotyping array and a “0” indicates that it is not present on the
genotyping array.</p>
<hr />
</div>
</div>
<div id="snpclip" class="section level3">
<h3>SNPclip</h3>
<div id="function-9" class="section level4">
<h4>Function</h4>
<pre><code>SNPclip(snps, 
        pop = &quot;CEU&quot;,
        r2_threshold = &quot;0.1&quot;, 
        maf_threshold = &quot;0.01&quot;, 
        token = NULL, 
        file = FALSE,
        genome_build = &quot;grch37&quot;
       )</code></pre>
<p>Prune a list of variants by linkage disequilibrium. Input is a list
of variant RS numbers (one per line) and a population group.</p>
</div>
<div id="arguments-9" class="section level4">
<h4>Arguments</h4>
<ul>
<li><code>snps</code>, a list of between 1 - 5,000 variants, using an RS
number (rsID) or chromosome coordinate (GRCh37) (e.g. “chr7:24966446”).
All input variants must be on the same chromosome and match a bi-allelic
variant.</li>
<li><code>pop</code>, a 1000 Genomes Project reference population, uses
three letter population code, (e.g. YRI or CEU), multiple allowed,
default = “CEU”</li>
<li><code>r2_threshold</code>, Used to set the R<sup>2</sup> threshold
for LD pruning. One of each pair of variants with a R<sup>2</sup>
greater than the threshold is removed. Value needs to be in the range 0
to 1. Default value is 0.1.</li>
<li><code>maf_threshold</code>, Used to set minor allele frequency (MAF)
threshold for LD pruning. Variants with a MAF less than or equal to the
threshold are removed. Value needs to be in the range 0 to 1. Default
value is 0.01.</li>
<li><code>token</code>, LDlink provided user access token is required,
default = NULL, register for a free token on the <a href="https://ldlink.nih.gov/?tab=apiaccess">LDlink web site.</a></li>
<li><code>file</code>, optional character string naming a path and file
for saving results. If file = FALSE, no file will be generated, default
= FALSE.</li>
<li><code>genome_build</code> Choose between one of three
options…<code>grch37</code> for genome build GRCh37 (hg19),
<code>grch38</code> for GRCh38 (hg38), or
<code>grch38_high_coverage</code> for GRCh38 High Coverage (hg38) 1000
Genome Project data sets. Default is GRCh37 (hg19).</li>
</ul>
</div>
<div id="usage-multiple-variants" class="section level4">
<h4>Usage: Multiple Variants</h4>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">SNPclip</span>(<span class="at">snps =</span>  <span class="fu">c</span>(<span class="st">&quot;rs3&quot;</span>, <span class="st">&quot;rs4&quot;</span>, <span class="st">&quot;rs148890987&quot;</span>, <span class="st">&quot;rs115955931&quot;</span>), </span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">pop =</span> <span class="st">&quot;YRI&quot;</span>, </span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">r2_threshold =</span>  <span class="st">&quot;0.1&quot;</span>, </span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">maf_threshold =</span> <span class="st">&quot;0.01&quot;</span>, </span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">token =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;LDLINK_TOKEN&quot;</span>),</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">genome_build =</span> <span class="st">&quot;grch37&quot;</span></span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>       )</span></code></pre></div>
<pre><code>##     RS_Number       Position          Alleles
## 1         rs3 chr13:32446842 C=0.829, T=0.171
## 2         rs4 chr13:32447222 A=0.829, G=0.171
## 3 rs148890987 chr13:32403784     C=1.0, T=0.0
## 4 rs115955931 chr13:32130008 G=0.954, A=0.046
##                                             Details
## 1                                     Variant kept.
## 2 Variant in LD with rs3 (R2=1.0), variant removed.
## 3              Variant MAF is 0.0, variant removed.
## 4                                     Variant kept.</code></pre>
<p>The output table provides details including query variant RS number,
genomic position, alleles, and and details about whether the variant was
kept or removed.</p>
<hr />
</div>
</div>
</div>
<div id="utilities-and-examples" class="section level2">
<h2>Utilities and Examples</h2>
<div id="list_chips" class="section level3">
<h3>list_chips</h3>
<div id="function-10" class="section level4">
<h4>Function</h4>
<pre><code>list_chips()</code></pre>
<p>Provides a data frame listing the names and abbreviation codes for
available commercial SNP Chip Arrays from Illumina and Affymetrix.</p>
</div>
<div id="usage-1" class="section level4">
<h4>Usage</h4>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list_chips</span>()</span></code></pre></div>
<hr />
</div>
</div>
<div id="list_pop" class="section level3">
<h3>list_pop</h3>
<div id="function-11" class="section level4">
<h4>Function</h4>
<pre><code>list_pop()</code></pre>
<p>Provides a data frame listing the available reference populations
from the 1000 Genomes Project, continental or super-populations
(e.g. European, African, Admixed American) and sub-populations (e.g
Finnish, Gambian, Peruvian)</p>
</div>
<div id="usage-2" class="section level4">
<h4>Usage</h4>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list_pop</span>()</span></code></pre></div>
<hr />
</div>
</div>
<div id="list_gtex_tissues" class="section level3">
<h3>list_gtex_tissues</h3>
<div id="function-12" class="section level4">
<h4>Function</h4>
<pre><code>list_gtex_tissues()</code></pre>
<p>Provides a data frame listing the GTEx full names,
<code>LDexpress</code> full names (without spaces) and acceptable
abbreviation codes of the 54 non-diseased tissue sites collected for the
<a href="https://gtexportal.org/home/">GTEx Portal</a> and used as input
for the <code>LDexpress</code> function.</p>
</div>
<div id="usage-3" class="section level4">
<h4>Usage</h4>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list_gtex_tissues</span>()</span></code></pre></div>
<hr />
</div>
</div>
</div>
<div id="faqs-frequently-asked-questions" class="section level2">
<h2>FAQs (Frequently Asked Questions)</h2>
<ol style="list-style-type: decimal">
<li><p>What if my access token doesn’t work?</p>
<ul>
<li>Please double check that the token was typed accurately. Then,
ensure the format of the function call is correct. For example, if your
alphanumeric access token is: 123abc456789, then, use it as:<br />
</li>
</ul>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">LDproxy</span>(<span class="at">snp =</span> <span class="st">&quot;rs456&quot;</span>, <span class="at">pop =</span> <span class="st">&quot;YRI&quot;</span>, <span class="at">token =</span> <span class="st">&quot;123abc456789&quot;</span>)</span></code></pre></div></li>
</ol>
<p><br></p>
<div class="line-block">           If you still can not solve the
problem, please email us at <a href="mailto:NCILDlinkWebAdmin@mail.nih.gov" class="email">NCILDlinkWebAdmin@mail.nih.gov</a>.</div>
<p><br></p>
<ol start="2" style="list-style-type: decimal">
<li><p>Can I set a threshold or cut-off value for R<sup>2</sup> or D`
values?</p>
<ul>
<li>No. <em>LDlinkR</em> functions do not include ‘threshold’ as an
argument. However, the returned data object can be subset using base R.
For example:</li>
</ul></li>
</ol>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">LDproxy</span>(<span class="st">&quot;rs12027135&quot;</span>, <span class="at">pop =</span> <span class="st">&quot;CEU&quot;</span>,<span class="at">r2d =</span> <span class="st">&quot;r2&quot;</span>, <span class="at">token =</span> <span class="st">&quot;YourTokenHere123&quot;</span>)</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>new_df <span class="ot">&lt;-</span> <span class="fu">subset</span>(df, R2 <span class="sc">&gt;=</span> <span class="fl">0.8</span>)</span></code></pre></div>
<p><br></p>
<ol start="3" style="list-style-type: decimal">
<li>I need to upload hundreds of variants from a text file into
LDmatrix. Why do I get an error with the following code?</li>
</ol>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;variant_list.txt&quot;</span>, <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a><span class="fu">LDmatrix</span>(<span class="at">snps =</span> test, <span class="at">pop =</span> <span class="st">&quot;CEU&quot;</span>, <span class="at">r2d =</span> <span class="st">&quot;r2&quot;</span>, <span class="at">token =</span> <span class="st">&quot;YourTokenHere123&quot;</span>)</span></code></pre></div>
<p><code>Error in LDmatrix(snps = test, pop = &quot;CEU&quot;, r2d = &quot;r2&quot;, token = &quot;YourTokenHere123&quot;),  :    Input is between 2 to 1000 variants.</code></p>
<p><br></p>
<ul>
<li>In the above example, ‘test’ is an object of class data frame and
the variants are in the first column. The first column needs to be
specified in order to avoid the error. The example below will yield the
desired result:</li>
</ul>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">&quot;variant_list.txt&quot;</span>, <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="fu">LDmatrix</span>(<span class="at">snps =</span> test[,<span class="dv">1</span>], <span class="at">pop =</span> <span class="st">&quot;CEU&quot;</span>, <span class="at">r2d =</span> <span class="st">&quot;r2&quot;</span>, <span class="at">token =</span> <span class="st">&quot;YourTokenHere123&quot;</span>)</span></code></pre></div>
<pre><code>##    RS_number rs60676332 rs7805287 rs127 rs456 rs10239961 rs114
## 1 rs60676332      1.000     0.008 0.013 0.017      0.286 0.039
## 2  rs7805287      0.008     1.000 0.980 0.882      0.170 0.614
## 3      rs127      0.013     0.980 1.000 0.900      0.167 0.632
## 4      rs456      0.017     0.882 0.900 1.000      0.177 0.722
## 5 rs10239961      0.286     0.170 0.167 0.177      1.000 0.008
## 6      rs114      0.039     0.614 0.632 0.722      0.008 1.000</code></pre>
<p><br></p>
<ol start="4" style="list-style-type: decimal">
<li><p>What genome build does LDlink use for genomic coordinates?</p>
<ul>
<li>All genomic coordinates are based on GRCh37/hg19.</li>
</ul></li>
</ol>
<p><br></p>
<ol start="5" style="list-style-type: decimal">
<li><p>How can I ask for help?</p>
<ul>
<li>If you find that you can’t answer a question or solve a problem
yourself, you can email us at <a href="mailto:NCILDlinkWebAdmin@mail.nih.gov" class="email">NCILDlinkWebAdmin@mail.nih.gov</a> or open an issue at the
<em>LDlinkR</em> GitHub repository (<a href="https://github.com/CBIIT/LDlinkR" class="uri">https://github.com/CBIIT/LDlinkR</a>).</li>
</ul></li>
</ol>
</div>
<div id="session-information" class="section level2">
<h2>Session Information</h2>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.2.3 (2023-03-15)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur ... 10.16
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## loaded via a namespace (and not attached):
##  [1] digest_0.6.31   R6_2.5.1        jsonlite_1.8.4  evaluate_0.20   cachem_1.0.7    rlang_1.0.6    
##  [7] cli_3.6.0       rstudioapi_0.14 jquerylib_0.1.4 bslib_0.4.2     rmarkdown_2.20  tools_4.2.3    
## [13] xfun_0.37       yaml_2.3.7      fastmap_1.1.1   compiler_4.2.3  htmltools_0.5.4 knitr_1.42     
## [19] sass_0.4.5</code></pre>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
