---
title: 'LDlinkR: an R package for rapid query of LDlink web-based linkage disequilibrium
  applications'
author: "Timothy A. Myers, Stephen J. Chanock and Mitchell J. Machiela"
date: "30 Jul 2019"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{LDlinkR: an R package for rapid query of LDlink web-based linkage disequilibrium applications}
  %\VignetteEngine{knitr::rmarkdown}
  \usagepackage[utf8]{inputenc}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("LDlinkR")
```

***

## Description

LDlink is an interactive and powerful suite of web-based tools for querying germline variants in population groups of interest to generate interactive tables and plots.

*LDlinkR* is an R package developed to query and download results generated by LDlink web-based applications from the R console.  *LDlinkR* accelerates genomic research by providing efficient and user-friendly functions to programmatically interrogate pairwise linkage disequilibrium from large lists of variants. 

## Installation
- The stable version of *LDlinkR* can be installed from CRAN:
```{r eval=FALSE}
install.packages("LDlinkR")
```
- The development version of the *LDlinkR* package can be installed from the GitHub repository by using the *devtools* package:
```{r eval=FALSE}
devtools::install_github("CBIIT/LDlinkR")
```
*LDlinkR* depends on the following package:

- *utils*, version 3.5.2 or later
- *httr*, version 1.4.0 or later

## Personal Access Token - Required

In order to access the LDlink API via *LDlinkR*, we use a personal access token. This is a common convention followed by many APIs and emulates the more familiar HTTPS username/password or SSH keys.

You will need to:

- Make a one-time request for your personal access token from a web browser at <https://ldlink.nci.nih.gov/?tab=apiaccess>.
- Once registered, your personal access token will be emailed to you. It is a string of 12 random letters and numbers. 
- Provide it as an argument when using *LDlinkR* (See examples below).

(Optional) The best security practice is to store your personal access token as an envionment variable where *LDlinkR* can find it and use it on your behalf but where it will not be accidentally shared with the public.  Step-by-step instructions follow:

After retrieving your personal access token from your email, put your token in your `.Renviron` file. `.Renviron` is a hidden file that lives in your home directory.  The easiest way to both find and edit the `.Renviron` file is with a function from the *usethis* package.  From the R console, do:

```{r eval=FALSE}
usethis::edit_r_environ()
```
Your `.Renviron` file should open in your editor.  Add a line that looks like this:
```{r eval=FALSE}
LDLINKR_TOKEN=yourtokenhere123
```
**Important**, ensure you put a line break at the end by hitting the *enter/return* key.

Save and close the `.Renviron` file.  Restart R, as environment variables are only loaded from `.Renviron` at the start of a new R session. Now, check to see that your token is available by entering:
```{r}
Sys.getenv("LDLINKR_TOKEN")
```
You should see your personal access token print to the screen, as shown above. Now, *LDlinkR* function calls that use
```{r eval=FALSE}
Sys.getenv("LDLINKR_TOKEN")
```
for the `token` argment in *LDlinkR* function calls will use your personal access token in a private and secure way.  This method will be used in the extended examples that follow.

***

## Functions

### LDhap

#### Function 
```
LDhap(snps, pop="CEU", token=NULL, file = FALSE)
```
Calculates population specific haplotype frequencies of all haplotypes observed for a list of query variants. Input is a list of variant RS numbers (concatenated list) and a population group.

#### Arguments
- `snps`, a list of between 1 - 30 variants, using an rsID or chromosome coordinate (e.g. "chr7:24966446")
- `pop`, a 1000 Genomes Project population, uses three letter population code, (e.g. YRI or CEU), multiple allowed, default = "CEU"
- `token`, LDlink provided user access token is required, default = NULL
- `file`, optional character string naming a path and file for saving results. If file = FALSE, no file will be generated, default = FALSE

#### Usage: Multiple query variants, single population
```{r}
LDhap(snps = c("rs3", "rs4", "rs148890987"), 
      pop = "CEU", 
      token = Sys.getenv("LDLINK_TOKEN")
     )
```
<br>

#### Usage: Multiple query variants, multiple populations
```{r}
LDhap(snps = c("rs3", "rs4", "rs148890987"),
      pop = c("YRI", "CEU"),
      token = Sys.getenv("LDLINK_TOKEN")
     )
```

Output is a table of alleles, haplotype count and haplotype frequencies. 

***

### LDmatrix

#### Function
```
LDmatrix(snps, pop = "CEU", r2d = "r2", token = NULL, file = FALSE)
```
Generates a data frame of pairwise linkage disequilibirum statistics. Input is a list of between 2 to 1000 variants. Desired output can be based on estimates of R^2^ or D'.

#### Arguments
- `snps`, list of between 2 - 1,000 variants, using an rsID or chromosome coordinate (GRCh37/hg19) (e.g. "chr7:24966446")
- `pop`, a 1000 Genomes Project population, uses three letter population code, (e.g. YRI or CEU), multiple allowed, default = "CEU"
- `r2d`, use either "r2" for pairwise R^2^ statistics or "d" for pairwise D' statistics
- `token`, LDlink provided user access token is required, default = NULL
- `file`, optional character string naming a path and file for saving results. If file = FALSE, no file will be generated, default = FALSE

#### Usage: Multiple query variants, single population, R^2^
```{r}
LDmatrix(snps = c("rs496202", "rs11147477", "rs201578600"), 
         pop = "YRI", r2d = "r2", 
         token = Sys.getenv("LDLINK_TOKEN")
        )
```
<br>

#### Usage: Multiple query variants (rsID & genomic coordinates), multiple populations, D'
```{r}
LDmatrix(snps = c("chr13:32444611", "rs11147477", "rs201578600"), 
         pop = c("YRI", "CEU"), r2d = "d", 
         token = Sys.getenv("LDLINK_TOKEN")
        )
```

Output is a table with rows and columns equal to the number of query variants and pairwise linkage disequilibrium statistics.

***

### LDpair

#### Function
```
LDpair(var1, var2, pop = "CEU", token = NULL, output = "table", file = FALSE)
```
Investigates potentially correlated alleles for a pair of variants.  Input is two query variants and a 1000 Genomes Project reference population(s) of interest.

#### Arguments
- `var1`, the first RS number (rsID) or genomic coordinate (GRCh37/hg19) (e.g. "chr7:24966446"), must match a bi-allelic variant
- `var2`, the second RS number or genomic coordinate, as above, must match a bi-allelic variant
- `pop`, a 1000 Genomes Project reference population, uses three letter population code, (e.g. YRI or CEU), multiple allowed, default = "CEU"
- `token`, LDlink provided user access token is required, default = NULL
- `output`, two output format options are available, "text", which displays a two-by-two matrix displaying haplotype counts and allele frequencies along with other statistics, or "table", which displays the same data in rows and columns, default = "table"
- `file`, optional character string naming a path and file for saving results. If file = FALSE, no file will be generated, default = FALSE

#### Usage: With `output` argument set to "text"
```{r}
LDpair(var1 = "rs496202", 
       var2 = "rs11147477", 
       pop = "YRI", 
       token = Sys.getenv("LDLINK_TOKEN"), 
       output = "text"
      )
```
<br>

#### Usage: With no `output` argument option specified, using default "table"
```{r}
LDpair(var1 = "rs496202", 
       var2 = "rs11147477", 
       pop = "YRI", 
       token = Sys.getenv("LDLINK_TOKEN")
      )
```
Output of the `output` argument "text" option is a two-by-two contingency table displaying haplotype counts and allele frequencies of the two query variants. Also displayed are calculated metrics of linkage disequilibrium including: D prime (D'), R square (R^2^), and goodness-of-fit (Chi-square and p-value). Goodness-of-fit tests for deviations of expected haplotype frequencies based on allele frequencies. Correlated alleles are reported if linkage disequilibrium is present (R^2^ > 0.1). If linkage equilibrium, no alleles are reported. 

Output from the `output` argument "table" option converts the data from the two-by-two contingency table into a data frame.

***

### LDpop

#### Function
```
LDpop(var1, var2, pop = "CEU", r2d = "r2", token = NULL, file = FALSE)
```
Investigates allele frequencies and linkage disequilibrium patterns across 1000G populations.

#### Arguments
- `var1`, the first RS number (rsID) or genomic coordinate (GRCh37/hg19) (e.g. "chr7:24966446"), must match a bi-allelic variant
- `var2`, the second RS number or genomic coordinate, as above, must match a bi-allelic variant
- `pop`, a 1000 Genomes Project reference population, uses three letter population code, (e.g. YRI or CEU), multiple allowed, default = "CEU"
- `r2d`, use "r2" if desired output is based on estimated R^2^ or "d" if D'
- `token`, LDlink provided user access token is required, default = NULL
- `file`, optional character string naming a path and file for saving results. If file = FALSE, no file will be generated, default = FALSE

#### Usage
```{r}
LDpop(var1 = "rs496202", 
      var2 = "rs11147477", 
      pop = "YRI", 
      r2d = "r2", 
      token = Sys.getenv("LDLINK_TOKEN")
     )
```

***

### LDproxy

#### Function
```
LDproxy(snp, pop = "CEU", r2d = "r2", token = NULL, file = FALSE)
```
Explore proxy and putatively functional variants for a single query variant. Input is a single RS number and a population group.  Depending on the number of query populations, this function could take some time to run.

#### Arguments
- `snp`, an RS number (rsID) or chromosome coordinate (GRCh37/hg19) (e.g. "chr7:24966446"), one per query, RS number must match a bi-allelic variant 
- `pop`, a 1000 Genomes Project reference population, uses three letter population code, (e.g. YRI or CEU), multiple allowed, default = "CEU"
- `r2d`, use "r2" if desired output is based on estimated R^2^ or "d" if D'
- `token`, LDlink provided user access token is required, default = NULL
- `file`, optional character string naming a path and file for saving results. If file = FALSE, no file will be generated, default = FALSE

#### Usage: single reference population
```{r}
my_proxies <- LDproxy(snp = "rs456", 
                      pop = "YRI", 
                      r2d = "r2", 
                      token = Sys.getenv("LDLINK_TOKEN")
                     )
```

Output is a data frame stored in the variable `my_proxies` with 2455 rows and 10 columns with data. 
```{r}
head(my_proxies)
```
Includes information on all variants -/+ 500 Kb of the query variant with a pairwise R^2^ value greater than 0.01.

***

### LDproxy_batch

#### Function
```
LDproxy_batch(snp, pop = "CEU", r2d = "r2", token = NULL, append = FALSE)
```
Query LDproxy using a list of query variants. *LDproxy_batch* will make sequential queries, one query per variant.  Concurrent queries are not permitted by the LDlink API.  Ouptput is saved as text file(s) to the current working directory. Depending on the number of query variants and reference populations selected, this function could time some time to run.

#### Arguments
- `snp`, a character string or data frame listing RS numbers (rsID) or chromosome coordinates (GRCh37/hg19) (e.g. "chr7:24966446"), one per line.
- `pop`, a 1000 Genomes Project reference population, uses three letter population code, (e.g. YRI or CEU), multiple allowed, default = "CEU"
- `r2d`, use "r2" if desired output is based on estimated R^2^ or "d" if D'
- `token`, LDlink provided user access token is required, default = NULL
- `append`, a logical, if TRUE, output for eqch query variant is appended to a single text file and saved to the current working directory.  If FALSE, output for each query variant is saved in its own text file with the query variant as the filename.  Default value is FALSE.

#### Usage: multiple variants, default `pop` and `r2d` 
The list of query variants passed to *LDproxy_batch* can be stored as a character string.
```{r eval=FALSE}
LDproxy_batch(snp = c("rs456", "rs114", "rs127"), 
              token = Sys.getenv("LDLINK_TOKEN")
             )
```
Or, a longer list of variants can be read into a data frame from a text file and passed into *LDproxy_batch*. The list should be in a simple text file, one query variant per line.  For example:
```{r}
my_variants <- read.table("variant_list.txt")
my_variants
```
Then, call *LDproxy_batch* with:
```{r eval=FALSE}
LDproxy_batch(snp = my_variants, 
              token = Sys.getenv("LDLINK_TOKEN")
             )
```
Output not displayed.  All output from *LDproxy_batch* is saved to a text file(s) in the current working directory.

***

### SNPchip

#### Function
```
SNPchip(snps, chip = "ALL", token = NULL, file = FALSE)
```
Used to find commercial genotyping chip arrays for variants.  Input is a list of between 1 - 5000 variants (one per line) and desired commercial chip arrays to search. Input variants do not need to be on the same chromosome. 

#### Arguments
- `snps`, between 1 - 5,000 variants, using an rsID or chromosome coordinate (e.g. "chr7:24966446")
- `chip`, chip or arrays, platform code(s) for a SNP chip array, ALL_Illumina, ALL_Affy or ALL, default=ALL, use the `list_chips` utility (see below) to lookup available commercial SNP chip arrays and their codes.
- `token`, LDlink provided user access token is required is required, default = NULL
- `file`, optional character string naming a path and file for saving results. If file = FALSE, no file will be generated, default = FALSE

#### Usage: Multiple variants, search "ALL" available chip arrays
```{r}
SNPchip(snps = c("rs3", "rs4", "rs148890987"), 
        chip = "ALL", 
        token = Sys.getenv("LDLINK_TOKEN")
       )
```

<br>

#### Usage: Multiple variants, search two Affymetrix arrays
```{r}
SNPchip(snps = c("rs3", "rs4", "rs148890987"), 
        chip = c("A_SNP5.0", "A_CHB2"), 
        token = Sys.getenv("LDLINK_TOKEN")
       )
```

<br>

#### Usage: Multiple variants, search all available Affymetrix arrays using, "ALL_Affy"
```{r}
SNPchip(snps = c("rs3", "rs4", "rs148890987"), 
        chip = "ALL_Affy", 
        token = Sys.getenv("LDLINK_TOKEN")
       )
```
Output is a data frame of query variant rows (RS number), genomic coordinate (GRCh37) and genotyping chip array columns. The presence of a "1" designates the variant is present on the respective commercial genotyping array and a "0" indicates that it is not present on the genotyping array. 

***

### SNPclip

#### Function
```
SNPclip(snps, pop = "CEU", r2_threshold = "0.1", maf_threshold = "0.01", token = NULL, file = FALSE)
```
Prune a list of variants by linkage disequilibrium. Input is a list of variant RS numbers (one per line) and a population group.

#### Arguments
- `snps`, a list of between 1 - 5,000 variants, using an RS number (rsID) or chromosome coordinate (GRCh37) (e.g. "chr7:24966446"). All input variants must be on the same chromosome and match a bi-allelic variant.
- `pop`, a 1000 Genomes Project reference population, uses three letter population code, (e.g. YRI or CEU), multiple allowed, default = "CEU"
- `r2_threshold`, Used to set the R^2^ threshold for LD pruning. One of each pair of variants with a R^2^ greater than the threshold is removed. Value needs to be in the range 0 to 1. Default value is 0.1.
- `maf_threshold`, Used to set minor allele frequency (MAF) threshold for LD pruning. Variants with a MAF less than or equal to the threshold are removed. Value needs to be in the range 0 to 1. Default value is 0.01.
- `token`, LDlink provided user access token is required is required, default = NULL
- `file`, optional character string naming a path and file for saving results. If file = FALSE, no file will be generated, default = FALSE

#### Usage:  Multiple Variants
```{r}
SNPclip(snps =  c("rs3", "rs4", "rs148890987", "rs115955931"), 
        pop = "YRI", 
        r2_threshold =  "0.1", 
        maf_threshold = "0.01", 
        token = Sys.getenv("LDLINK_TOKEN")
       )
```
The output table provides details including query varaint RS number, genomic position, alleles, and and details about whether the variant was kept or removed.

***

## Utilities

### list_chips

#### Function
```
list_chips()
```
Provides a data frame listing the names and abbreviation codes for available commerical SNP Chip Arrays from Illumina and Affymetrix.

#### Usage
```{r}
list_chips()
```

***

### list_pop

#### Function
```
list_pop()
```
Provides a data frame listing the available reference populations from the 1000 Genomes Project, continetal or super-populations (e.g. European, African, Admixed American) and sub-populations (e.g Finnish, Gambian, Peruvian)

#### Usage
```{r}
list_pop()
```

***

## Session Information
```{r}
sessionInfo()
```






